src/Servers/Kestrel/Core/src/Adapter/Internal/AdaptedPipeline.cs

```diff
@@ -44,8 +45,16 @@ public async Task RunAsync(Stream stream)
             var inputTask = ReadInputAsync(stream);
             var outputTask = WriteOutputAsync(stream);
 
-            await inputTask;
             await outputTask;
+            // Yield the current request
+            _transport.Input.CancelPendingRead();
```

Comment:
Shouldn't we wait for the connection middleware pipeline before closing the input? I think this breaks half-duplex if that matters.